### PR TITLE
Add just the test infrastructure bits from #4354

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

#4354 is a pretty complicated PR. It's got a bunch of conpty changes, but what it also has was some critical improvements to the roundtrip test suite. I'm working on some other bugfixes in the same area currently, and need these tests enhancements in those branches _now_. The rest of #4354 is complex enough that I don't trust it will get merged soon (if ever). However, these fixes _should_ be in regardless.

## PR Checklist
* [x] Taken directly from #4354
* [x] I work here
* [x] Tests added/passed
* [n/a] Requires documentation to be updated

## Detailed Description of the Pull Request / Additional comments

This is four main changes:
* Enable conpty to be fully enabled in unittests. Just setting up a VT renderer isn't enough to trick the host into being in conpty mode - it also needs to have some other flags set.
* Some minor changes to `CommonState` to better configure the common test state for conpty
* Move some of the verify helpers from `ConptyRoundtripTests` into their own helper class, to be shared in multiple tests
* Add a `TerminalBufferTests` class, for testing the Terminal buffer directly (without conpty).

This change is really easier than 
image
would suggest, I promise.

### CODE REVIEW

**Comment by miniksa** [comment]
If this succeeds, it might just be the solution to #4384

**Comment by miniksa** [comment]
OK it succeeded twice. I think this might solve the x64 unit test failures then. Reading it fully now. @DHowett-MSFT

File: src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp [hunk 1]
```diff
@@ -144,18 +169,27 @@ class ConptyRoundtripTests
...
+                               NoThrowString().Format(L"writing=\"%hs\", expecting %u strings", actualString.c_str(), expectedOutput.size()));
 
-    Log::Comment(NoThrowString().Format(L"Expected =\t\"%hs\"", first.c_str()));
-    Log::Comment(NoThrowString().Format(L"Actual =\t\"%hs\"", actualString.c_str()));
+        std::string first = expectedOutput.front();
+        expectedOutput.pop_front();
 
-    VERIFY_ARE_EQUAL(first.length(), cch);
-    VERIFY_ARE_EQUAL(first, actualString);
+        Log::Comment(NoThrowString().Format(L"Expected =\t\"%hs\"", first.c_str()));
```
Comment (DHowett-MSFT) [inline src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp]: will this print raw escapes, or escape them?

Comment (DHowett-MSFT) [reply src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp]: should we replace `0x1b` with `0x241b`?

**Comment by zadjii-msft** [comment]
For the record, the manual run I did above was just to give us more data if this does in fact fix #4344. The tests did pass, so it seems to be looking good. I don't know what I did, but if this next run passes too, lets just close #4344 as well.

**Comment by zadjii-msft** [comment]
WELP

/azp run

**Comment by zadjii-msft** [comment]
Well https://dev.azure.com/ms/Terminal/_build/results?buildId=60592&view=logs&j=e7365288-57d7-5349-8dfa-9c195127bd77 is curious.

The x86 test pass already has errors with `UiaTextRangeTests::CanMoveByWord_NonEmptyBuffer` and `UiaTextRangeTests::CanMoveEndpointByUnitWord`. But it kept on chugging along till it got to the roundtrip/output tests, where it got stuck again, in x86 this time though. x64 already passed.

**Pull request merged**